src/librustdoc/html/render/mod.rs

```diff
@@ -1305,15 +1288,46 @@ fn render_impl(
             }
         };
 
-        let (is_hidden, extra_class) =
-            if (trait_.is_none() || item.doc_value().is_some() || item.kind.is_type_alias())
-                && !is_default_item
-            {
-                (false, "")
-            } else {
-                (true, " hidden")
-            };
         let in_trait_class = if trait_.is_some() { " trait-impl" } else { "" };
+
+        let mut doc_buffer = Buffer::empty_from(w);
+        let mut info_buffer = Buffer::empty_from(w);
+        let mut short_documented = true;
+
+        if render_method_item {
+            if !is_default_item {
+                if let Some(t) = trait_ {
+                    // The trait item may have been stripped so we might not
+                    // find any documentation or stability for it.
+                    if let Some(it) = t.items.iter().find(|i| i.name == item.name) {
+                        // We need the stability of the item from the trait
+                        // because impls can't have a stability.
+                        if item.doc_value().is_some() {
+                            document_item_info(&mut info_buffer, cx, it, Some(parent));
+                            document_full(&mut doc_buffer, item, cx);
+                            short_documented = false;
+                        } else {
+                            // In case the item isn't documented,
+                            // provide short documentation from the trait.
+                            document_short(&mut doc_buffer, it, cx, link, parent, show_def_docs);
+                        }
+                    }
+                } else {
+                    document_item_info(&mut info_buffer, cx, item, Some(parent));
+                    if show_def_docs {
+                        document_full(&mut doc_buffer, item, cx);
+                        short_documented = false;
+                    }
+                }
+            } else {
+                document_short(&mut doc_buffer, item, cx, link, parent, show_def_docs);
+            }
+        }
+        let tmp_w = if short_documented && trait_.is_some() { short_w } else { w };
```

Comment:
The name `tmp_w` doesn't tell me much about what this variable does (and indeed I actually don't know what it does). Can we name it something more descriptive?

Reply:
I actually have no idea how to rename it better... Originally, I simply had in mind to use variable shadowing by naming it `w`, but that seemed a bit counter-intuitive when I read the code afterwards, so I simply went for `tmp-w`. If you have better suggestions, I'm all ear!

Reply:
Sometimes when I'm stuck for variable names, it helps to come up with some terms that are very evocative even if not necessarily technical, and then define them technically.

So here, I would change the fn definition:

```
    // For trait implementations, the `interesting` output contains all methods that have
    // doccomments, and the `boring` output contains all methods that do not. The
    // distinction is used to allow hiding the boring methods.
    fn doc_impl_item(
        boring: &mut Buffer,
        interesting: &mut Buffer,
        cx: &Context<'_>,
        item: &clean::Item,
        parent: &clean::Item,
```

Then down here you would do:

```
        let w = if short_documented && trait_.is_some() { interesting } else { boring };
```

Reply:
Fine by me!